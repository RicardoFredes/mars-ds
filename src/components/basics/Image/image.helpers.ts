import type { ImageItem } from "./image.types";

type Emulation = "none" | "dark" | "light";

export const mockPrefersColorScheme = (emulation: Emulation) => {
  const getMatch = (emulation: Emulation, query: string) =>
    query === `(prefers-color-scheme: ${emulation})`;

  Object.defineProperty(window, "matchMedia", {
    writable: true,
    value: jest.fn().mockImplementation((query) => ({
      matches: getMatch(emulation, query),
      media: query,
      onchange: null,
    })),
  });
};

type ParseImageSourcesProps = Partial<{
  src: string;
  darkModeSrc: string;
  forceDarkMode: boolean;
  sm: ImageItem;
  md: ImageItem;
  lg: ImageItem;
}>;

export const parseImageSources = ({
  src,
  darkModeSrc,
  forceDarkMode,
  sm,
  md,
  lg,
}: ParseImageSourcesProps) => {
  const defaultSrc = src || sm?.src || md?.src || lg?.src;
  const defaultDarkModeSrc = darkModeSrc || sm?.darkModeSrc || md?.darkModeSrc || lg?.darkModeSrc;
  if (forceDarkMode) {
    return {
      xs: defaultDarkModeSrc || defaultSrc,
      sm: sm?.darkModeSrc || sm?.src,
      md: md?.darkModeSrc || md?.src,
      lg: lg?.darkModeSrc || lg?.src,
    };
  }
  return {
    xs: defaultSrc || defaultDarkModeSrc,
    sm: sm?.src || sm?.darkModeSrc,
    md: md?.src || md?.darkModeSrc,
    lg: lg?.src || lg?.darkModeSrc,
  };
};
