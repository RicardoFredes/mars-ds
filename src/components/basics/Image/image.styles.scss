.image {
  display: block;
}

.theme-dark {
  .image {
    &--auto-invert-color {
      filter: invert(1);
    }
  }
}
