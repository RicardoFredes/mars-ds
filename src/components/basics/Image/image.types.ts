import type { HTMLAttributes } from "react";

interface ImageSourceProps extends ImageItem {
  media?: string;
  mediaSize?: number;
}

interface ImageItem {
  src?: string;
  darkModeSrc?: string;
  height?: string | number;
  width?: string | number;
}

interface ImageItemProps extends ImageItem {
  sm?: ImageItem;
  md?: ImageItem;
  lg?: ImageItem;
}

interface ImageProps extends HTMLAttributes<HTMLImageElement>, ImageItemProps {
  autoDarkMode?: boolean;
  maxHeight?: string | number;
}

export type { ImageProps, ImageSourceProps };
