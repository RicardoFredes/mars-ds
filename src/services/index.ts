export * from "./compare";
export * from "./makeArray";
export * from "./masker";
export * from "./names-parser";
export * from "./validate";
