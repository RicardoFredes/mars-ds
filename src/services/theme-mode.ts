type ColorScheme = "dark" | "light" | "auto";

const THEME_KEY = "theme";

const isColorSchemaForced = (colorScheme: ColorScheme) => {
  if (typeof window === "undefined" || typeof window.matchMedia === "undefined") return false;
  try {
    const themeName = getThemeValue(colorScheme);
    const localValue = localStorage.getItem(THEME_KEY) === themeName;
    const systemValue = matchMedia(`(prefers-color-scheme: ${colorScheme})`).matches;
    if (colorScheme === "dark" && window.localStorage.getItem("theme") === "theme-light") {
      return false;
    }
    return localValue !== systemValue;
  } catch (error) {
    console.error(error);
    return false;
  }
};

const getThemeValue = (themeName: ColorScheme) => `theme-${themeName}`;

export const isDarkModeForced = () => isColorSchemaForced("dark");
export const isLightModeForced = () => isColorSchemaForced("light");

export const setTheme = (theme?: ColorScheme) => {
  if (!theme) localStorage.removeItem(THEME_KEY);
  const colorSchema = matchMedia("(prefers-color-scheme: dark)").matches ? "dark" : "light";
  const themeName = getThemeValue(theme || colorSchema);
  localStorage.setItem(THEME_KEY, themeName);
  document.body.classList.add(themeName);
};
