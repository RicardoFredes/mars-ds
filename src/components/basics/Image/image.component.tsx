import type { ImageProps, ImageSourceProps } from "./image.types";

import classNames from "classnames";
import { useEffect, useMemo, useState } from "react";

import { isDarkModeForced, isLightModeForced } from "@/services/theme-mode";

import Skeleton from "../Skeleton";
import { parseImageSources } from "./image.helpers";

const Image = ({
  autoDarkMode,
  className,
  src,
  height,
  width,
  sm,
  md,
  lg,
  darkModeSrc,
  title,
  maxHeight,
  alt,
  ...props
}: ImageProps) => {
  const [forceDarkMode, setForceDarkMode] = useState(false);
  const [forceLightMode, setForceLightMode] = useState(false);
  const [ready, setReady] = useState(false);

  useEffect(() => {
    if (isDarkModeForced()) {
      setForceDarkMode(true);
    } else {
      setForceLightMode(isLightModeForced());
    }
    setReady(true);
  }, []);

  const images = useMemo(
    () => parseImageSources({ src, darkModeSrc, forceDarkMode, sm, md, lg }),
    [darkModeSrc, forceDarkMode, lg, md, sm, src]
  );
  if (Object.values(images).every((value) => !value)) return null;

  const defaultHeight = height || sm?.height || md?.height || lg?.height;
  const defaultWidth = width || sm?.width || md?.width || lg?.width;

  if (!ready) return <Skeleton active width={defaultWidth} height={defaultHeight} />;

  const style = { maxHeight };
  const commonProps = { height: defaultHeight, width: defaultWidth, style };
  const cn = classNames("image", className, { "image--auto-invert-color": autoDarkMode });

  return (
    <picture data-testid="image" className={cn} {...props}>
      {!forceLightMode && darkModeSrc && (
        <source media={"(prefers-color-scheme: dark)"} srcSet={darkModeSrc} />
      )}
      {images.sm && (
        <ImageSource mediaSize={300} {...sm} src={images.sm} disableDarkMode={forceLightMode} />
      )}
      {images.md && (
        <ImageSource mediaSize={640} {...md} src={images.md} disableDarkMode={forceLightMode} />
      )}
      {images.lg && (
        <ImageSource mediaSize={920} {...lg} src={images.lg} disableDarkMode={forceLightMode} />
      )}
      <img data-testid="img" src={images.xs} alt={alt || title} {...commonProps} />
    </picture>
  );
};

const ImageSource = ({
  darkModeSrc,
  src,
  height,
  width,
  media,
  mediaSize,
  disableDarkMode,
}: ImageSourceProps) => {
  const m = media || `(min-width: ${mediaSize}px)`;
  const style = { maxHeight: height };
  const commonProps = { height, width, style };

  if (disableDarkMode || !darkModeSrc) {
    return <source media={m} srcSet={src} {...commonProps} />;
  }

  const md = media || `${m} and (prefers-color-scheme: dark)`;
  return (
    <>
      <source media={m} srcSet={src} {...commonProps} />
      {<source media={md} srcSet={darkModeSrc} {...commonProps} />}
    </>
  );
};

export default Image;
