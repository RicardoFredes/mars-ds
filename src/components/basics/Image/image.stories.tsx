import type { ComponentMeta, ComponentStory } from "@storybook/react";

import { commonStoreComponentProps } from "@/services/stories";

import Image from "./image.component";

const myStoryFunctionDescription =
  'This Image component was designed to support advanced uses. You can just use it as if it were an "img" tag or pass more specific arguments to define different images in the dark theme. Furthermore it is possible to specify different arguments for different screen sizes.';

const commonArgs = {
  width: {
    description: "The image width",
    control: "number",
    type: { name: "number", required: false },
  },
  height: {
    description: "The image height",
    control: "number",
  },
  src: {
    description: "The default url source image",
    control: "text",
  },
  darkModeSrc: {
    description: "The url source image for darkmode theme",
    control: "text",
  },
};

const getResolutionDescription = (resolution: string) => {
  const args = Object.keys(commonArgs).join(", ");
  return `The above "${resolution}" resolution attributes for: ${args}`;
};

export default {
  title: "Components/Image",
  parameters: {
    docs: {
      storyDescription: myStoryFunctionDescription,
    },
    loki: { skip: true },
  },
  component: Image,
  argTypes: {
    title: {
      description: "The title and alt attribute image",
      control: "text",
    },
    autoDarkMode: {
      description: "Invert this image when it is true",
      control: "boolean",
      defaultValue: false,
    },
    maxHeight: {
      description: "The image max-height style",
      control: "number",
    },
    ...commonArgs,
    sm: {
      description: getResolutionDescription("sm"),
      control: "object",
      ...commonArgs,
    },
    md: {
      description: getResolutionDescription("md"),
      control: "object",
      ...commonArgs,
    },
    lg: {
      description: getResolutionDescription("lg"),
      control: "object",
      ...commonArgs,
    },
    ...commonStoreComponentProps,
  },
} as ComponentMeta<typeof Image>;

const Template: ComponentStory<typeof Image> = (args) => <Image {...args} />;

const commonImageArgs = {
  width: undefined,
  height: undefined,
  src: undefined,
  darkModeSrc: undefined,
};

const DefaultArgs = {
  title: "Imagem de exemplo",
  width: 438.7,
  height: 400,
  src: "https://cdn.mesalva.com/uploads/image/MjAyMi0wNS0yNyAxNzo1NTozMiArMDAwMDYzODU5Mw%3D%3D%0A.png",
  darkModeSrc:
    "https://cdn.mesalva.com/uploads/image/MjAyMi0wNS0yNyAxNzo1NTozMiArMDAwMDYzODU5Mw%3D%3D%0A.png",
};

export const Default = Template.bind({});
Default.args = DefaultArgs;

export const ScreenSizes = Template.bind({});
ScreenSizes.args = {
  ...DefaultArgs,
  sm: commonImageArgs,
  md: commonImageArgs,
  lg: commonImageArgs,
};
