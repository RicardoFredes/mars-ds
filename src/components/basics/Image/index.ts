/**
 * WARNING!
 * Do not change this file,
 * because it is automatically generated
 * run "yarn reindex" to update
 */

import Image from "./image.component";
export default Image;
export * from "./image.types";
