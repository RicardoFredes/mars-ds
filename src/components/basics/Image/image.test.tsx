import type { ImageProps } from "./image.types";

import { render } from "@testing-library/react";

import { isDarkModeForced, isLightModeForced } from "@/services/theme-mode";

import Image from "./image.component";
import { mockPrefersColorScheme, parseImageSources } from "./image.helpers";

const defaultProps = {
  title: "any_title",
};

const makeSut = (props?: ImageProps) => render(<Image {...defaultProps} {...props} />);

describe("<Image>", () => {
  it("should render component with basic props", () => {
    const wrapper = makeSut({
      src: "default.png",
      height: 150,
      width: 300,
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("should put specific className if autoDarkMode is true", () => {
    const wrapper = makeSut({ src: "default.png", autoDarkMode: true });
    const element = wrapper.getByTestId("image");
    expect(element).toHaveAttribute("class", "image image--auto-invert-color");
  });

  describe("when prefers colors schema is undefined", () => {
    beforeEach(() => mockPrefersColorScheme("none"));

    it("should be default src", () => {
      const src = "default.png";
      const wrapper = makeSut({ src });
      const image = wrapper.getByAltText(defaultProps.title);
      expect(image).toHaveAttribute("src", src);
    });

    describe("and has more sizes", () => {
      it("should be default src", () => {
        const src = "default.png";
        const sm = { src: "sm.png" };
        const wrapper = makeSut({ src, sm });
        const image = wrapper.getByAltText(defaultProps.title);
        expect(image).toHaveAttribute("src", src);
      });

      it("should not be default src", () => {
        const sm = { src: "sm.png" };
        const wrapper = makeSut({ sm });
        const image = wrapper.getByAltText(defaultProps.title);
        expect(image).toHaveAttribute("src", sm.src);
      });

      it("should render component with advanced props", () => {
        const wrapper = makeSut({
          src: "default.png",
          height: 150,
          width: 300,
          darkModeSrc: "darkmode.png",
          md: { src: "md.png" },
        });
        expect(wrapper).toMatchSnapshot();
      });
    });
  });

  describe("helpers", () => {
    describe("isDarkModeForced", () => {
      beforeEach(() => {
        window.localStorage.removeItem("theme");
      });
      it('when prefers-color-scheme is "dark" and localStorage is "theme-dark" should return false', () => {
        mockPrefersColorScheme("dark");
        window.localStorage.setItem("theme", "theme-dark");
        expect(isDarkModeForced()).toBe(false);
      });

      it('when prefers-color-scheme is "dark" and localStorage is "theme-light" should return false', () => {
        mockPrefersColorScheme("dark");
        window.localStorage.setItem("theme", "theme-light");
        expect(isDarkModeForced()).toBe(false);
      });

      it('when prefers-color-scheme is "dark" and localStorage isn\'t set should return true', () => {
        mockPrefersColorScheme("dark");
        expect(isDarkModeForced()).toBe(true);
      });
    });

    describe("isLightModeForced", () => {
      beforeEach(() => {
        window.localStorage.removeItem("theme");
      });
      it('when prefers-color-scheme is "light" and localStorage is "theme-dark" should return true', () => {
        mockPrefersColorScheme("light");
        window.localStorage.setItem("theme", "theme-dark");
        expect(isLightModeForced()).toBe(true);
      });

      it('when prefers-color-scheme is "light" and localStorage is "theme-light" should return false', () => {
        mockPrefersColorScheme("light");
        window.localStorage.setItem("theme", "theme-light");
        expect(isLightModeForced()).toBe(false);
      });

      it('when prefers-color-scheme is "light" and localStorage isn\'t set should return true', () => {
        mockPrefersColorScheme("light");
        expect(isLightModeForced()).toBe(true);
      });
    });

    describe("parseImageSources", () => {
      const sm = {
        src: "light_sm.jpg",
        darkModeSrc: "dark_sm.jpg",
        width: 150,
        height: 150,
        disableDarkMode: false,
      };

      const md = {
        src: "light_md.jpg",
        darkModeSrc: "dark_md.jpg",
        width: 400,
        height: 400,
        disableDarkMode: false,
      };

      const lg = {
        src: "light_lg.jpg",
        darkModeSrc: "dark_lg.jpg",
        width: 600,
        height: 600,
        disableDarkMode: false,
      };

      const args = {
        src: "light.jpg",
        darkModeSrc: "dark.jpg",
        forceDarkMode: false,
        sm,
        md,
        lg,
      };

      it("when forceDarkMode false should return an object with light images", () => {
        const obj = parseImageSources(args);
        expect(obj).toStrictEqual({
          md: "light_md.jpg",
          lg: "light_lg.jpg",
          sm: "light_sm.jpg",
          xs: "light.jpg",
        });
      });

      it("when forceDarkMode true should return an object with dark images", () => {
        const obj = parseImageSources({ ...args, forceDarkMode: true });
        expect(obj).toStrictEqual({
          xs: "dark.jpg",
          sm: "dark_sm.jpg",
          md: "dark_md.jpg",
          lg: "dark_lg.jpg",
        });
      });

      it("when forceDarkMode false should return an object with light images, if missing return the fallback", () => {
        const obj = parseImageSources({
          ...args,
          src: undefined,
          lg: { darkModeSrc: "dark_lg.jpg" },
        });
        expect(obj).toStrictEqual({
          md: "light_md.jpg",
          lg: "dark_lg.jpg",
          sm: "light_sm.jpg",
          xs: "light_sm.jpg",
        });
      });

      it("when forceDarkMode true should return an object with dark images, if missing return the fallback", () => {
        const obj = parseImageSources({
          ...args,
          forceDarkMode: true,
          darkModeSrc: undefined,
          md: { src: "light_md.jpg" },
        });
        expect(obj).toStrictEqual({
          xs: "dark_sm.jpg",
          sm: "dark_sm.jpg",
          md: "light_md.jpg",
          lg: "dark_lg.jpg",
        });
      });
    });
  });
});
