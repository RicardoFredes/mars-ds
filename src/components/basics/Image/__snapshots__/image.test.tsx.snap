// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image> should render component with basic props 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <picture
        class="image"
        data-testid="image"
      >
        <img
          alt="any_title"
          data-testid="img"
          height="150"
          src="default.png"
          width="300"
        />
      </picture>
    </div>
  </body>,
  "container": <div>
    <picture
      class="image"
      data-testid="image"
    >
      <img
        alt="any_title"
        data-testid="img"
        height="150"
        src="default.png"
        width="300"
      />
    </picture>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Image> when prefers colors schema is undefined and has more sizes should render component with advanced props 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <picture
        class="image"
        data-testid="image"
      >
        <source
          media="(prefers-color-scheme: dark)"
          srcset="darkmode.png"
        />
        <source
          media="(min-width: 640px)"
          srcset="md.png"
        />
        <img
          alt="any_title"
          data-testid="img"
          height="150"
          src="default.png"
          width="300"
        />
      </picture>
    </div>
  </body>,
  "container": <div>
    <picture
      class="image"
      data-testid="image"
    >
      <source
        media="(prefers-color-scheme: dark)"
        srcset="darkmode.png"
      />
      <source
        media="(min-width: 640px)"
        srcset="md.png"
      />
      <img
        alt="any_title"
        data-testid="img"
        height="150"
        src="default.png"
        width="300"
      />
    </picture>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
